src/hotspot/share/opto/node.hpp

```diff
@@ -2111,4 +2111,51 @@ inline int Op_DivModIL(BasicType bt, bool is_unsigned) {
   }
 }
 
+// Interface to define actions that should be taken when running DataNodeBFS. Each use can extend this class to specify
+// a customized BFS.
+class BFSActions : public StackObj {
+ public:
+  // Should a node's inputs further be visited in the BFS traversal? By default, we visit all data inputs. Override this
+  // method to provide a custom filter.
+  virtual bool should_visit(Node* node) const {
+    // By default, visit all inputs.
+    return true;
+  };
+
+  // Is the visited node a target node that we are looking for in the BFS traversal? We do not visit its inputs further
+  // but the BFS will continue to visit all unvisited nodes in the queue.
+  virtual bool is_target_node(Node* node) const = 0;
+
+  // Defines an action that should be taken when we visit a target node in the BFS traversal.
+  virtual void target_node_action(Node* target_node) = 0;
+};
+
+// Class to perform a BFS traversal on the data nodes from a given start node. The provided BFSActions guide which
+// data node's inputs should be further visited, which data nodes are target nodes and what to do with the target nodes.
+class DataNodeBFS : public StackObj {
+  BFSActions& _bfs_actions;
+
+ public:
+  explicit DataNodeBFS(BFSActions& bfs_action) : _bfs_actions(bfs_action) {}
```

Comment:
Is this restricted to data-nodes? If so, you should verify that the `start_node` is a data node. But we could also generalize this to any BFS, and then check in `should_visit` if it is a data node of CFG.

You should also say that it traverses inputs/def, not outputs/uses.

Reply:
Yes, it should only visit data nodes. Good point about generalizing it to any BFS, including CFG nodes. I gave it a shot (see new commit) but limited the patch to an input-only BFS (I guess it could be extended to an output-including or input/ouput-selecting BFS as well at some point but we should have a use case first).

For now, I'm only defining a `DataNodeInputsBFS` but provided a simple way to add a `CFGNodeInputsBFS` or a "any node" `NodeInputsBFS` later (again, I'm leaving these additional implementation out of this patch since we should first have a use case - but should be simple to add).

Let me know what you think.